New calibration data tree and onsite scripts

FrancaCassol opened the pull request:
This PR includes the following points:

1. It introduces a pixel calibration data tree ("/fefs/data/rea/monitoring/PixelCalibration") which will contain all the present pixel calibration files (both for drs4 and PMTs) 
2. It simplifies the onsite scripts used to fill this tree, so as to easily permit an automatic call of the scripts online 
3. It adds two new scripts:  one to reconstruct the filter scan data, the other to fit them in order to obtain the systematic noise to be used in the calibration procedure
4. It includes the systematic noise in the F-factor calibration formula

A schematic description of the data tree and of the scripts is given in this note: [CameraCalibrationNote.pdf](https://github.com/cta-observatory/cta-lstchain/files/6869019/CameraCalibrationNote.pdf)

maxnoe commented on lstchain/data/filters_trasmission.dat:
```diff
@@ -0,0 +1,37 @@
+63 0.00007
+64 0.00004
+65 0.00001
+66 0.00000
```
exactly 0? Or do we need more digits here?

FrancaCassol replied:
For the camera is practically zero, but I can indeed add some digit

maxnoe commented on lstchain/tools/lstchain_fit_intensity_scan.py:
```diff
@@ -240,6 +240,73 @@ def finish(self):
+                    for pix,i in enumerate([0,600,1200,1800]):
+                        pad+=1
+                        plt.subplot(4, 2, pad)
+                        plt.tight_layout()
```
Run tight layout once, just before saving the plot

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -25,6 +25,9 @@
                       type=int, nargs="+")
 
 version,subversion=lstchain.__version__.rsplit('.post',1)
+optional.add_argument('-f', '--filters_list', help="Filter list (same order than run list)",
```
same order as run list

FrancaCassol commented:
Hi,
I finalised this code with several improvements.

The code now works for creating the calibration data tree. 

For @morcuended:
You can find a tree example in 
/fefs/aswg/workspace/franca.cassol/data/real/monitoring/PixelCalibration 

For creating calibration files without systematic corrections (as it is now) one has to send the calibration script with the option 
--sys_date 0  

The output calibration file will have then the name "no_sys_corrected_calibration..." (I wanted to be sure not have confusion among the two cases). By default the script looks for the last available systematic corrections file before the run.

In order to test with this just copy my directory 
/fefs/aswg/workspace/franca.cassol/data/real/monitoring/PixelCalibration/ffactor_systematics
which contains correction files from three filter scans (for the moment I will take care for the production of these files)

Pay attention that in the case we apply the systematic corrections, the EventSource calibration scaling factors (calib_scale_high_gain, calib_scale_low_gain) must be changed to [1.088,1.004] in order to scale only for the integration window size.  I hope we can move soon to this way of proceeding

I updated the note [CameraCalibrationNote.pdf](https://github.com/cta-observatory/cta-lstchain/files/6945169/CameraCalibrationNote.pdf) and added more details on the scripts' logic.

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -0,0 +1,155 @@
+
+    if shutil.which('srun') is None:
+        print(">>> This script needs a slurm batch system. Stop")
+        return
```
`sys.exit(1)` to exit with an error code.

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -0,0 +1,155 @@
+
+    # verify config file
+    if not os.path.exists(config_file):
+        raise IOError(f"Config file {config_file} does not exists. \n")
```
Raising an exception in the main method does not really make sense. Print error and `sys.exit(2)`

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -0,0 +1,155 @@
+def main():
+
+    if shutil.which('srun') is None:
+        print(">>> This script needs a slurm batch system. Stop")
```
Use logging instead of prints

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -125,14 +126,15 @@ def main():
-                # job file
-                job_file = f"{log_dir}/{run}_{sub_run}.job"
+                # job file              
+                now = datetime.now().strftime("%d-%m-%Y-%H:%M:%S")
```
Please use ISO format `%Y-%m-%dT%H:%M:%S` or just 

```
In [7]: datetime.now().isoformat(sep='T')
Out[7]: '2021-09-15T15:25:39.910310'

In [8]: datetime.now().replace(microsecond=0).isoformat(sep='T')
Out[8]: '2021-09-15T15:25:43'
```

moralejo commented:
Thanks a lot for these developments. I went through the changes & new scripts an to me this seems to be in  a good shape for merging. But probably it is better to wait for green light from @morcuended after testing it on-site. @morcuende, did you have a chance to test it? This should simplify the calibration part of lstosa.

morcuended commented:
on it

morcuended reviewed (changes requested):
Apart from the small comments, the thing that should be changed in my opinion is how the lstchain version is obtained. Especially when you are working with a fixed tag, the sub-version will not be available and the script will fail.

Side note, I've not been able to produce ffactor_systematics files with `lstchain/scripts/onsite/onsite_create_ffactor_systematics_file.py` but I guess it is a matter of setting properly the configuration file.

FrancaCassol commented on lstchain/scripts/lstchain_data_create_drs4_pedestal_file.py:
```diff
@@ -66,8 +66,10 @@
 source_config = {
     "LSTEventSource": {
         "max_events":args.max_events,
+        "apply_drs4_corrections": False
```
It is false because here we want calculate the corrections so they must not be applied.  
There is a warning because this trailets concerns a next version of ctapipe_io_lst, but since it does not damage I added it already

maxnoe replied:
I just made the release of the new `ctapipe_io_lst` version, so you could update the requirement here

FrancaCassol replied:
Are you sure we want to do it here?

maxnoe replied:
If you set an option that is only valid for the new version, yes. If not, don't set the option and set it in the update PR

FrancaCassol replied:
Ok, I take it out

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_file.py:
```diff
@@ -22,40 +29,58 @@
+optional.add_argument('-p', '--pedestal_run', help="Pedestal run to be used. If None, it looks for the pedestal run of the date of the FF data.",
+                      type=none_or_str)
+
+version,subversion=lstchain.__version__.rsplit('.post',1)
```
Why remove the subversion at all?

morcuended replied:
subversion is not used anywhere in the code, it is just to unpack directly the version

maxnoe replied:
Yes, so the effective is that the subversion is not used for the path, it is removed. I Argumente tagt we don't want to do that.

morcuended replied:
Ah! Now I follow you. You propose to simply leave `version=lstchain.__version__` to make explicit that a non-tagged version is used. I agree with that.

maxnoe commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -0,0 +1,166 @@
+                    fh.write("#SBATCH --job-name=%s.job\n" % run)
+                    fh.write("#SBATCH --output=log/run_%s_subrun_%s_date_%s.out\n" % (run,sub_run,now))
+                    fh.write("#SBATCH --error=log/run_%s_subrun_%s_date_%s.err\n" % (run,sub_run,now))
+                    fh.write("#SBATCH -A dpps\n")
```
This should be an option for the script then probably

FrancaCassol replied:
Actually is this needed?

morcuended replied:
I do not think so. It will just assume your group account if not specified. The line can be removed.

FrancaCassol commented on lstchain/scripts/onsite/onsite_create_calibration_file.py:
```diff
@@ -69,61 +94,90 @@ def main():
         path, date = input_dir.rsplit('/', 1)
```
Hi @morcuended, 

sorry, I do not understand your comment, what do you mean?

morcuended replied:
Just a comment, nothing to be changed: the only way right now to know which date corresponds to which run number is by looking at the directory tree (what it is done here) or to look into the CaCo db.

FrancaCassol commented on lstchain/scripts/onsite/onsite_create_drs4_pedestal_file.py:
```diff
@@ -21,8 +22,9 @@
 
 required.add_argument('-r', '--run_number', help="Run number with drs4 pedestals",
                       type=int, required=True)
+version,subversion=lstchain.__version__.rsplit('.post',1)
```
Ops, I just modified in a different way, let me know if it is fine

morcuended commented on lstchain/scripts/onsite/onsite_create_calibration_file.py:
```diff
@@ -22,40 +31,77 @@
+optional.add_argument('-p', '--pedestal_run', help="Pedestal run to be used. If None, it looks for the pedestal run of the date of the FF data.",
+                      type=none_or_str)
+
+version=lstchain.__version__.rsplit('.post',1)[0]
```
What Max was proposing is to keep the `.post...` part explicitly.

maxnoe replied:
Yes, please just use `__version__` so that test / development tests cannot be confused with the normal productions.

FrancaCassol replied:
Sorry, but this is exactly what I do not want, because it demands to recreate a new calibration tree at each commit. 
I think this solution is good both for the official tree  than for the debugging (head) code

morcuended replied:
@FrancaCassol made me realize that it would be very difficult in the debugging phase because one should reproduce a new tree subdirectory at each commit.

maxnoe replied:
Then I would propose to just remove the commit part, not the post part.

FrancaCassol replied:
what does the post number mean?

morcuended replied:
I think it is the commit hash.

FrancaCassol replied:
So I do not like to have it in the tree either

maxnoe replied:
The part after `post` is the number of commits since the tag and then comes `git<hash>`.

I was proposing to keep the `post` part, not the part after that. So that any development version is e.g.
```
0.7.5.post
```

Note, that with the switch  to `setuptools_scm` in #737, this would be:

```
0.7.6.dev
```

FrancaCassol replied:
I think it is an unuseful complication for debugging. It does not concerns the official tree (which will be alway with a release), and developer people do not need to add a "post" to the personal tree to remember that they are working with a development release. Also it could be I like to use part of the official tree but if my tree has a different version name I can not do it.

morcuended replied:
I'm fine with the version line as it is

FrancaCassol commented on lstchain/scripts/onsite/onsite_create_drs4_time_file.py:
```diff
@@ -0,0 +1,129 @@
+if config_file is None:
+    config_file = os.path.join(os.path.dirname(__file__), "../../data/onsite_camera_calibration_param.json")
+
+max_events = 1000000
```
yes, it is not needed here, I remove it

FrancaCassol commented on lstchain/scripts/onsite/onsite_create_calibration_file.py:
```diff
@@ -69,100 +114,232 @@ def main():
+        log_file = f"{output_dir}/log/{output_name}.log"
+        print(f"\n--> Log file {log_file}")
+
+        if os.path.exists(output_file):
             if query_yes_no(">>> Output file exists already. Do you want to remove it?"):
                 os.remove(output_file)
+                os.remove(log_file)
```
Actually in slurm it gives an immediate error, I wonder, is there a way to know if a script is running in slurm?

morcuended replied:
You can use either `squeue` or `sacct` (with the `-j` if you know the job_id)

maxnoe replied:
The `srun` echos the job id, so you get it when starting the job. But this is again somethign that I would rather see in the onsite analysis than in lstchain, this goes rather deep into cluster management, not analysis.

FrancaCassol replied:
I found an other way, hope it is fine ;-)

FrancaCassol commented on lstchain/scripts/onsite/onsite_create_calibration_files_with_batch.py:
```diff
@@ -0,0 +1,166 @@
+                    fh.write("#SBATCH --mem-per-cpu=10G\n")
+                    fh.write("#SBATCH -D %s \n" % output_dir)
+
+                    cmd = f"srun onsite_create_calibration_file -r {run} " \
+                        f"-p {ped_run} -v {prod_id} --sub_run {sub_run} " \
+                        f"-b {base_dir} -s {stat_events} --output_base_name {output_base_name} " \
+                        f"--filters {filters} --sys_date {sys_date} " \
+                        f"--config {config_file} --time_run {time_run}"
+
+                    if no_sys_correction:
+                        cmd += " --no_sys_correction"
+
+                    fh.write(cmd)
```
Yes, it would be nice, do you perhaps know how to do it?

maxnoe replied:
The problem is that you run a batch command and the submission was successful, the job itself failed.

To keep track of which jobs failed and why, you need a much more complex system, that e.g. stores submitted jobs in a database and periodically checks if the job succeeded or not.

FrancaCassol replied:
yes, and with slurm is not so easy because the error files are always produced and some time with not meaningful errors

maxnoe replied:
You should make sure that stdout and stderr are going into the same file and that exit codes are correctly propagated in the scripts. E.g. bash scripts should always have `set -euo pipefail` at the start.

FrancaCassol replied:
why in the same file?

morcuended commented on lstchain/scripts/onsite/onsite_create_calibration_file.py:
```diff
@@ -235,16 +235,20 @@ def main():
+        yes = False
         if os.path.exists(output_file):
-            if query_yes_no(">>> Output file exists already. Do you want to remove it?"):
+            if os.getenv('SLURM_JOB_ID') is None:
```
I think this variable can be only interpreted inside a script launched with sbatch. Here it will be always None.

maxnoe replied:
indeed. This is the current johb id, not the one of a job launches using `sbatch` and friends

FrancaCassol replied:
yes, and it will be not None when the script is running in batch (this is the idea)

maxnoe replied:
But the check is to make sure that in SLURM, the script is not asking for user input.

However, there is a better way: adding the conventional `--yes` option that will prevent dialogs and answers them so that the script continues automatically.

FrancaCassol replied:
?

maxnoe replied:
It is convention that many programs follow, to add a `-y, --yes` command line option that will prevent any dialogs.

e.g. like this:

```
parser.add_argument('-y', '--yes', help='Do not ask interactively for permissions, assume yes')
```

and then for example later:

```
overwrite = args.yes or query_yes_no("Overwrite?")
```

maxnoe replied:
The sbatch script can then add the `--yes` option to prevent asking for any user input

FrancaCassol replied:
I see, never met this option. But the present implementation has the advantage to automatically switch off the dialog in case of batch (without recalling to use the option),  what is the problem with it?

morcuended replied:
Both ways are fine. Please note that there are other (onsite) scripts with the same problem though.

FrancaCassol replied:
right

FrancaCassol replied:
I added also the yes option, it can indeed be useful in case one want to directly overwrite in the interactive case

FrancaCassol commented:
Dear @maxnoe and @morcuended, 
what is missing to finalize this PR? 
I need to make further developments for the automatic monitoring of the calibration data and I would like to do it after this PR is approved. Thanks!

jsitarek commented on lstchain/tools/lstchain_fit_intensity_scan.py:
```diff
@@ -0,0 +1,364 @@
+
+    fractional_variance_error = Float(
+        0.02,
+        help='Costant fractional error assumed for the y fit coordinate (variance)'
```
Costant ==> Constant
However why do you use the constant uncertainty? You can compute it easily from the variance itself and number of values.

FrancaCassol replied:
I prefer to leave the possibility to add some systematic contribution. By the way, which formula would you use?

jsitarek replied:
sigma_error = sigma / sqrt(2N - 2), where N is the number of measurements from which sigma is computed. 
see e.g.: https://math.stackexchange.com/questions/1259383/calculating-uncertainty-in-standard-deviation
I tested it with a simple toy MC and it works perfect well for a simple gaussian distribution and > 10 measurements. 0.02 that you are using would correspond to ~1250 events. I guess this is more or less in the correct ballpark for the statistics that you are using, but if you have large differences in statistics between different filter wheel positions the fit might be more stable/precise if you use proper uncertainties

jsitarek commented on lstchain/tools/lstchain_fit_intensity_scan.py:
```diff
@@ -0,0 +1,364 @@
+        quadratic_term = np.ma.array(self.fit_parameters.T[1], mask=self.fit_error.T)
+
+        # give to the badly fitted pixel a median value for the B term
+        median_quadratic_term = np.ma.median(quadratic_term,axis=0)
```
I suggest here to add an extra check - if >~5-10% of the pixels have their fit wrong the whole script should end with an error.

FrancaCassol replied:
I prefer to added a plot of the fitted B values over the camera, this will give a clear idea of the pixels that were badly  fit (there are also the text warning)

jsitarek commented on lstchain/tools/lstchain_fit_intensity_scan.py:
```diff
@@ -0,0 +1,364 @@
+
+
+def quadratic_fit(t, b=1, c=1):
+    f2 = 1.222
```
it is not nice to have this F2 value hardcoded here (I think it only affects the gain, while you want to compute first of all the quadratic term, but still, the gain itself is also put in the output, so the value should be consistent with what is used everywhere else)

FrancaCassol replied:
yes, my problem was how to pass a constant parameter to the fitted function. I found a solution

FrancaCassol commented on lstchain/calib/camera/calibration_calculator.py:
```diff
@@ -152,21 +181,14 @@ def calculate_calibration_coefficients(self, event):
-        # find signal median of good pixels
+        # find signal median of good pixels (FF factor=<npe>/npe)
         masked_npe = np.ma.array(n_pe, mask=calib_data.unusable_pixels)
         npe_signal_median = np.ma.median(masked_npe, axis=1)
```
yes, I will add a comment

jsitarek commented on lstchain/calib/camera/calibration_calculator.py:
```diff
@@ -141,9 +158,21 @@ def calculate_calibration_coefficients(self, event):
+        gain = np.divide(numerator, denominator, out=np.zeros_like(numerator), where=denominator != 0)
+
+        # correct for the quadratic term if possible
+        if self.quadratic_term is not None:
```
as you prefer. B=0 I only suggested because then you can directly use one formula (that would not work with "none"), if you prefer to make it in two steps it can also leave as it is now
